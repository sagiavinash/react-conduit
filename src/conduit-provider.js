import React, { Component, Children } from "react";
import PropTypes from "prop-types";
import Registry from "./registry";

class ConduitProvider extends Component {
  constructor(props) {
    super(props);
    const { registry } = this.props;

    this.registry = registry ? registry : new Registry();
  }

  getChildContext() {
    return { registry: this.registry };
  }

  render() {
    return Children.only(this.props.children);
  }
}

ConduitProvider.childContextTypes = {
  registry: PropTypes.object.isRequired,
};

export default ConduitProvider;
