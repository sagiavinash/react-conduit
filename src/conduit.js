class Conduit {
  constructor(inlet, outlet) {
    this.inlet = inlet;
    this.outlet = outlet;
  }

  disconnect = () => {
    this.inlet.onDisconnect(this);
    this.outlet.onDisconnect(this);
  };

  connect = () => {
    this.inlet.onConnect(this);
    this.outlet.onConnect(this);
  };

  getId = () => `${this.inlet.getId()}-${this.outlet.getId()}`;
  getInlet = () => this.inlet;
  getOutlet = () => this.outlet;
  update = () => this.outlet.forceRender();
  members = () => ({ inlet: this.inlet, outlet: this.outlet });
  toString = () => `${this.getId()}: ${this.inlet.getLabel()} -> ${this.outlet.getLabel()}`;
}

export default Conduit;
