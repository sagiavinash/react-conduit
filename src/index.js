export { default as ConduitProvider } from "./conduit-provider";
export { default as Inlet } from "./inlet";
export { default as Outlet } from "./outlet";
export { default as Registry } from "./registry";
