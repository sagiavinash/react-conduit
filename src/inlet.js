import React, { Component, Children } from "react";
import PropTypes from "prop-types";
import uuidV4 from "uuid/v4";
import { updateChildren, removeInlet, addInlet } from "./registry";

export class Inlet extends Component {
  constructor(props) {
    super(props);
    this.id = uuidV4();
  }

  componentDidUpdate(prevProps) {
    if (this.props.label !== prevProps.label) {
      this.context.registry.rewireInlet(this);
    }
    this.context.registry.updateConduits(this);
  }

  componentWillMount() {
    this.context.registry.registerInlet(this);
  }

  componentWillUnmount() {
    this.context.registry.unregisterInlet(this);
  }

  // Registry Inlet API
  getId = () => this.id;
  getLabel = () => this.props.label;
  getIndex = () => this.props.index;
  getChildren = () => this.props.children;
  onDisconnect = conduit => {
    this.props.onDisconnect(conduit.simplify());
  };
  onConnect = conduit => {
    this.props.onConnect(conduit.simplify());
  };

  render() {
    return null;
  }
}

Inlet.contextTypes = {
  registry: PropTypes.object.isRequired,
};

Inlet.propTypes = {
  label: PropTypes.string.isRequired,
  index: PropTypes.number,
  onDisconnect: PropTypes.func,
  onConnect: PropTypes.func,
};

Inlet.defaultProps = {
  index: 0,
  onDisconnect: () => {},
  onConnect: () => {},
};

export default Inlet;
