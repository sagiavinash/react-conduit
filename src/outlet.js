import React, { Component } from "react";
import PropTypes from "prop-types";
import uuidV4 from "uuid";

class Outlet extends Component {
  constructor(props, context) {
    super(props, context);

    this.id = uuidV4();
    this.state = { children: this.getChildren() };
  }

  componentDidUpdate(prevProps) {
    if (this.props.label !== prevProps.label) {
      this.context.registry.rewireOutlet(this);
    }
  }

  componentDidMount() {
    this.context.registry.registerOutlet(this);
  }

  componentWillUnmount() {
    this.context.registry.unregisterOutlet(this);
  }

  getChildren = () => this.context.registry.mergeChildrenForLabel(this.getLabel());

  // Registry Outlet API
  getId = () => this.id;
  getLabel = () => this.props.label;
  onDisconnect = conduit => {
    this.props.onDisconnect(conduit.members());
  };
  onConnect = conduit => {
    this.props.onConnect(conduit.members());
  };
  forceRender = () => this.setState({ children: this.getChildren() });

  render() {
    return (
      <div className={this.props.className} style={this.props.style}>
        {this.state.children}
      </div>
    );
  }
}

Outlet.contextTypes = {
  registry: PropTypes.object.isRequired,
};

Outlet.propTypes = {
  label: PropTypes.string.isRequired,
  className: PropTypes.string,
  style: PropTypes.object,
  onDisconnect: PropTypes.func,
  onConnect: PropTypes.func,
};

Outlet.defaultProps = {
  className: null,
  style: null,
  onDisconnect: () => {},
  onConnect: () => {},
};

export default Outlet;
