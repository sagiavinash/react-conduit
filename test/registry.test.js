import React from "react";
import { mount, render } from "enzyme";
import { ConduitProvider, Inlet, Outlet, Registry } from "../src";

const mountWithRegister = (element, registry) => mount(element, { context: { registry } });
const mountWithProvider = element =>
  mount(
    <ConduitProvider>
      {element}
    </ConduitProvider>,
  );

describe("registry", () => {
  let registry;

  beforeEach(() => {
    registry = new Registry();
  });

  it("should merge all the children from matching inlets", () => {
    const wrapper = render(
      <ConduitProvider>
        <div>
          <Inlet label="1">
            <div>c1</div>
          </Inlet>
          <Inlet label="1">
            <div>c2</div>
          </Inlet>
          <Inlet label="1" index={-1}>
            <div>c0</div>
          </Inlet>
          <Outlet label="1" />
        </div>
      </ConduitProvider>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("should update outlet when inlets change their children", () => {
    const inlet = mountWithRegister(
      <Inlet label="1">
        <p>content</p>
      </Inlet>,
      registry,
    );
    const outlet = mountWithRegister(<Outlet label="1" />, registry);

    expect(outlet.html()).toEqual("<div><p>content</p></div>");
    inlet.setProps({ children: <p>new content</p> });
    expect(outlet.html()).toEqual("<div><p>new content</p></div>");
  });

  it("should update outlets when inlets change their labels", () => {
    const inlet = mountWithRegister(
      <Inlet label="1">
        <p>content</p>
      </Inlet>,
      registry,
    );
    const outlet1 = mountWithRegister(<Outlet label="1" />, registry);
    const outlet2 = mountWithRegister(<Outlet label="2" />, registry);

    expect(outlet1.html()).toEqual("<div><p>content</p></div>");
    expect(outlet2.html()).toEqual("<div></div>");
    inlet.setProps({ label: "2" });
    expect(outlet1.html()).toEqual("<div></div>");
    expect(outlet2.html()).toEqual("<div><p>content</p></div>");
  });

  it("should update outlet when it changes its label", () => {
    const inlet1 = mountWithRegister(
      <Inlet label="1">
        <p>content 1</p>
      </Inlet>,
      registry,
    );
    const inlet2 = mountWithRegister(
      <Inlet label="2">
        <p>content 2</p>
      </Inlet>,
      registry,
    );
    const outlet = mountWithRegister(<Outlet label="1" />, registry);

    expect(outlet.html()).toEqual("<div><p>content 1</p></div>");
    outlet.setProps({ label: "2" });
    expect(outlet.html()).toEqual("<div><p>content 2</p></div>");
  });

  it("should update outlet when a new inlet is rendered", () => {
    const inlet1 = mountWithRegister(
      <Inlet label="1">
        <p>content 1</p>
      </Inlet>,
      registry,
    );
    const outlet = mountWithRegister(<Outlet label="1" />, registry);

    expect(outlet.html()).toEqual("<div><p>content 1</p></div>");
    mountWithRegister(
      <Inlet label="1">
        <p>content 2</p>
      </Inlet>,
      registry,
    );
    expect(outlet.html()).toEqual("<div><p>content 1</p><p>content 2</p></div>");
  });

  it("should update outlet when an inlet is unmounted", () => {
    const inlet1 = mountWithRegister(
      <Inlet label="1">
        <p>content 1</p>
      </Inlet>,
      registry,
    );
    const inlet2 = mountWithRegister(
      <Inlet label="1">
        <p>content 2</p>
      </Inlet>,
      registry,
    );
    const outlet = mountWithRegister(<Outlet label="1" />, registry);

    expect(outlet.html()).toEqual("<div><p>content 1</p><p>content 2</p></div>");
    inlet2.unmount();
    expect(outlet.html()).toEqual("<div><p>content 1</p></div>");
  });

  it("should invoke onDisconnect on both inlet and outlet when a conduit gets disconnected", () => {
    const inletSpy = jest.fn();
    const outletSpy = jest.fn();

    const inlet = mountWithRegister(
      <Inlet label="1" onDisconnect={inletSpy}>
        <p>content</p>
      </Inlet>,
      registry,
    );
    const outlet = mountWithRegister(<Outlet label="1" onDisconnect={outletSpy} />, registry);

    expect(outlet.html()).toEqual("<div><p>content</p></div>");
    inlet.unmount();
    expect(inletSpy).toHaveBeenCalledTimes(1);
    expect(outletSpy).toHaveBeenCalledTimes(1);
  });

  it("should invoke onConnect on both inlet and outlet when a new conduit is connected", () => {
    const inletSpy = jest.fn();
    const outletSpy = jest.fn();

    const inlet = mountWithRegister(
      <Inlet label="1" onConnect={inletSpy}>
        <p>content</p>
      </Inlet>,
      registry,
    );
    const outlet = mountWithRegister(<Outlet label="1" onConnect={outletSpy} />, registry);

    expect(inletSpy).toHaveBeenCalledTimes(1);
    expect(outletSpy).toHaveBeenCalledTimes(1);
  });
});
