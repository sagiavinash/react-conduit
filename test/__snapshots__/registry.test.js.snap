// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registry should merge all the children from matching inlets 1`] = `
<div>
  <div>
    <div>
      c0
    </div>
    <div>
      c1
    </div>
    <div>
      c2
    </div>
  </div>
</div>
`;
